dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -25,52 +25,88 @@
 import java.util.List;
 import java.util.Objects;
 import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.graph.Dependency;
 
 /**
- * Path from the root to a node in a dependency tree, where a node is a dependency.
+ * Path from the root to a node in a dependency tree, where the root node is a Maven artifact and
+ * non-root nodes are dependencies.
+ *
+ * <p>The root node is null for the dependency trees generated for multiple artifacts by {@link
+ * DependencyGraphBuilder#buildFullDependencyGraph(List)}; otherwise the root node is not null.
  */
 public final class DependencyPath {
 
-  private List<Dependency> path = new ArrayList<>();
+  // The root of the dependency path. The project root is not a dependency.
+  private final Artifact root;
+  // Path without the root
+  private final List<Dependency> path = new ArrayList<>();
+
+  public DependencyPath(@Nullable Artifact root) {
+    this.root = root;
+  }
 
   @VisibleForTesting
-  public void add(Dependency dependency) {
-    path.add(dependency);
+  public DependencyPath appended(Dependency dependency) {
+    DependencyPath copy = new DependencyPath(root);
+    for (Dependency value : path) {
```

Comment:
I see a lot of copying of lists. It looks like appending a single element copies every item in the list, and then appending another element copies every item in the list again. This is probably N^2 performance and might be relevant given the size of some of our trees.

Reply:
levelOrder was already copying the list when building the dependency tree that consists of `DependencyPath`s
https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1371/files#diff-659d7bf4148f942fcc1eb60a50a2ee80L275 .

In future, we can abandon `DependencyPath` in favor of our own node class to represent a tree. It would be something similar to Maven's DependencyNode, but with a parent field. Maven's DependencyNode does not know its parent; therefore, given a DependnecyNode, we cannot get the path from the root to the node.

Reply:
You might do better by copying/cloning the entire list rather than by rebuilding it.